PR : 실습 2주차 과제

Jokwanhee opened the pull request:
## 📌𝘐𝘴𝘴𝘶𝘦𝘴
- #5 

## 📎𝘞𝘰𝘳𝘬 𝘋𝘦𝘴𝘤𝘳𝘪𝘱𝘵𝘪𝘰𝘯
- **바텀 네비게이션**
안드로이드, 홈, 마이페이지

- **리사이클러뷰 구현**
멀티 뷰 타입 지원  (sealed class 사용)
^ 헤더 : 나의 프로필 (리스트 맨 처음에 추가)
^ 콘텐트 : 생일인 친구 (친구 중 생일을 체크하여 자료구조 큐를 사용하여 첫 번째 인덱스에 차례로 추가)
^ 홈 : 친구   
 
- **업데이트 추가**
데이터 스키마에 update 파라미터를 추가하여 하루 이내에 갱신된 내용이 있으면 프로필 이미지 왼쪽상단에 표시

- **제약조건을 활용한 elipsize 설정하기**
constriantLayout 적극 활용하여 프로필 이름, 상태 메시지, 상태 바에 대응하여 글자 컨텐츠 마지막 ... 표현

## 📷𝘚𝘤𝘳𝘦𝘦𝘯𝘴𝘩𝘰𝘵

https://github.com/DO-SOPT-ANDROID/kwanhee-jo/assets/90740783/cc364637-47bd-4540-a47e-3a1c8dd6ff47


## 💬𝘛𝘰 𝘙𝘦𝘷𝘪𝘦𝘸𝘦𝘳𝘴
이번 과제는 구현은 잘 된 것 같은데, 설계할 때 고민을 많이했습니다..
더더더더더 잘할 수 있을 것 같은데, 뭔가 아쉬운 부분이 많아요 ㅠ (코드를 더 간소화할 수 있지 않을까?)
리뷰 부탁드립니다! 👍

Jokwanhee committed:
[add] fragment and bottomnavigation setting

프래그먼트 및 바텀내비게이션 셋팅

Jokwanhee committed:
[add] fragment and bottomNavigation

아.. 커밋 잘못했다. 여기가 진짜 프래그먼트 및 바텀내비게이션 셋팅..

Jokwanhee committed:
[add] recyclerview impl

Jokwanhee committed:
[add] multiviewtype and UI binding

리사이클러뷰 멀티 뷰 타입

Marchbreeze commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeAdapter.kt:
```diff
@@ -0,0 +1,75 @@
+
+    fun setProfileList(profileList: List<HomeProfileModel>) {
+        this.profileList = profileList.toList()
+        notifyDataSetChanged()
```
요 친구 아래에 밑줄 그어지는거 거슬리시지 않나요? ㅎㅎ 심화과제에서 ListAdapter를 활용한다면 밑줄 없이 깔끔하게 진행하실 수 있습니닷 ~

librarywon replied:
ListAdapter쓰면 getItemCount함수도 없어지지롱

Jokwanhee replied:
오잉.. 거슬렸는데 빨리 도전과제하면서 ListAdapter 사용해야겠네요!

librarywon reviewed (commented):
멀티뷰타입 미쳤다 그냥 와우

Marchbreeze commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeContentsViewHolder.kt:
```diff
@@ -0,0 +1,32 @@
+        binding.tvUserDescription.text = item.description
+
+        if (item.profileImage == null) {
+            Glide.with(binding.root)
+                .load(R.drawable.kakao_profile)
+                .override(150)
+                .transform(MultiTransformation(CenterCrop(), RoundedCorners(50)))
```
Glide 라이브러리 활용 너무 좋네요 !! 다만 요즘은 대세가 [Coil 라이브러리](https://velog.io/@jojo_devstory/Android-%EC%9D%B4%EB%AF%B8%EC%A7%80-%EB%A1%9C%EB%94%A9-%EB%9D%BC%EC%9D%B4%EB%B8%8C%EB%9F%AC%EB%A6%AC-Coil-%EC%9D%84-%EC%95%8C%EC%95%84%EB%B3%B4%EC%9E%90) 활용으로 넘어가는 추세라 요 라이브러리에 대해서도 한번 알아보면 좋을 것 같아요 ㅎㅎ

Jokwanhee replied:
Coil.. 이 녀석도 한 번 사용해봐야겠네요!

Marchbreeze commented on app/src/main/java/org/sopt/dosopttemplate/presentation/home/HomeActivity.kt:
```diff
@@ -0,0 +1,71 @@
+class HomeActivity : BaseActivity<ActivityHomeBinding>(ActivityHomeBinding::inflate) {
+    private var backPressedTime = 0L
+
+    override fun initView() {
+        initFragment()
+    }
+
+    private fun initFragment() {
+        val currentFragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
+        if (currentFragment == null) {
+            replaceFragment(HomeFragment.newInstance())
+            binding.bottomNavigationView.selectedItemId = R.id.home
+        }
+    }
```
요 친구들을 따로 함수화해둔 이유가 있나요?

Jokwanhee replied:
initView는 액티비티 추상클래스의 추상함수여서 구현부를 따로 작성해야하는데.. 프래그먼트를 붙이는 구현부를 또 따로 함수로 빼볼까 하면서 함수를 만들었습니다. 따로 이유는 없습니다! 🤣

Marchbreeze commented on app/src/main/java/org/sopt/dosopttemplate/presentation/home/HomeFragment.kt:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private fun initList() {
+        testDeque.apply {
```
임시리스트로 담아둔 코드들을 따로 파일에 리스트로 저장해두고, Fragment 파일에서는 해당 리스트를 받아와서 활용하는 방식을 사용하신다면 더 깔끔하고 유지보수가 용이한 코드를 작성하실 수 있을 것 같습니다 ~

librarywon replied:
뷰모델을 사용한다면 더 깔끔한코드가 될 것 같네요

Jokwanhee replied:
@Marchbreeze 
[daef0fe](https://github.com/DO-SOPT-ANDROID/kwanhee-jo/pull/6/commits/daef0fe350fa326460d21c6628e2427133dee64c)

임시 리스트 파일을 만들어서 접근하니 훨씬 코드의 양이 적어지네요! 따로 분리하는 샘플 자료를 만든다는 것 굉장히 좋은 것 같습니다. 감사합니다~!

Marchbreeze reviewed (commented):
너무 잘하셔서... 코드리뷰 달만한게 없네요.... 미쳤다 이남자

librarywon commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeContentsViewHolder.kt:
```diff
@@ -0,0 +1,32 @@
+            Glide.with(binding.root)
+                .load(item.profileImage)
+                .override(150)
+                .transform(MultiTransformation(CenterCrop(), RoundedCorners(50)))
```
이건 원으로 잘라주는 친구인가보군요 글라이드 많이 안써봐서 진짜 모름

Jokwanhee replied:
맞습니당! 👋

librarywon commented on app/src/main/java/org/sopt/dosopttemplate/presentation/home/MyPageFragment.kt:
```diff
@@ -0,0 +1,50 @@
+
+    private fun initMyData() {
+        with(activity?.getSharedPreferenceUser()) {
+            binding.tvIdContent.text = this?.getString(ID, "") ?: return@with
```
`?: return@with` 이 문법은 뭔가요?

Jokwanhee replied:
엘비스 연산자를 사용해서 널일 때, with 스코프 함수를 나가게 만들었습니다.

librarywon commented on app/src/main/res/drawable/jokwanhee.jpg:
잘생기기까지 했다 이 남자

Jokwanhee replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 🤣 감사합니다!

librarywon commented on app/src/main/res/layout/fragment_mypage.xml:
```diff
@@ -0,0 +1,106 @@
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/tv_mbti_content"
+            app:layout_constraintVertical_bias="0.9" />
```
아마 실수로 들어갔을 듯 싶은데 지워주세요!

Jokwanhee replied:
요거 constraint 제약을 걸어서 오른쪽 하단에 로그아웃 버튼을 해둔건데 없애면 왼쪽 상단으로 이동합니다. 혹시 지워야할 부분이 다른 부분인가요!?

librarywon replied:
`layout_constraintVertical_bias` 이 코드가 필요한 상태인가요?

Jokwanhee replied:
저는 로그아웃 버튼을 더 오른쪽 하단에 깔리게 하기위해서 bias를 주었습니다! 근데 지금 보니깐 딱히 필요없어 보이네요.. 하.. 제가 원래 어렸을 때 그림을 참 못그렸는 데 여기서 또 이러네요; 감사합니다 💛 
[39f5e5f](https://github.com/DO-SOPT-ANDROID/kwanhee-jo/pull/6/commits/39f5e5fe30f0e19da24dc626ec5676aacd3d8834)

librarywon commented on app/src/main/res/layout/item_profile_contents.xml:
```diff
@@ -0,0 +1,107 @@
+        android:id="@+id/tv_status_message"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:background="@drawable/bg_ffffff_stroke_dd4f4f_2_corners_50"
```
오우...파일명 보소

haeti-dev commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeAdapter.kt:
```diff
@@ -0,0 +1,75 @@
+                    LayoutInflater.from(parent.context),
+                    parent, false
+                )
+            )
```
뷰타입을 Sealed Class를 활용해서 사용해봅시당

Jokwanhee replied:
[b196360](https://github.com/DO-SOPT-ANDROID/kwanhee-jo/pull/6/commits/b196360d9ca3c38669d01e8ec081937ebe077230)

멀티 뷰 타입을 사용해봤습니다! 훨씬 확장성 있게 작성된 것 같아요!

haeti-dev reviewed (commented):
오비분들이 코리를 많이 달아주시겠지만, 정말 좋은 코드들이 많네요. 자꾸 이러면.. 과제를 어렵게 내야하나.. ㅎ 
고생하셨습니다 최고최고 👍

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeHeaderViewHolder.kt:
```diff
@@ -0,0 +1,32 @@
+        binding.tvUserName.text = item.name
+        binding.tvUserDescription.text = item.description
+
+        if (item.profileImage == null) {
+            Glide.with(binding.root)
+                .load(R.drawable.kakao_profile)
+                .override(200)
+                .transform(MultiTransformation(CenterCrop(), RoundedCorners(50)))
+                .into(binding.ivUser)
+        } else {
+            Glide.with(binding.root)
+                .load(item.profileImage)
+                .override(200)
+                .transform(MultiTransformation(CenterCrop(), RoundedCorners(50)))
+                .into(binding.ivUser)
+        }
```
Glide 반복되는 로직이 많은데 확장이나 Util로 분리하면 좋을 것 같습니다~!!

Jokwanhee replied:
[4b45cfc](https://github.com/DO-SOPT-ANDROID/kwanhee-jo/pull/6/commits/4b45cfcc24893fe125137020fdaa9f50eac93847)

확장함수 사용하니 거의 12줄의 코드가 2줄로 줄었습니다! 가독성 측면에서 매우 좋은 것 같습니다 감사합니다 _ _)

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/model/HomeProfileModel.kt:
```diff
@@ -0,0 +1,3 @@
+package org.sopt.dosopttemplate.model
+
+sealed class HomeProfileModel
```
어떠한 의도로 모델파일들을 분리해놓으셨나요?!

Jokwanhee replied:
리스트의 타입을 하나로 묶기 위해서 사용했습니다. 리스트를 받고 각 요소들이 어떤 데이터 클래스를 가지는 가에 따라서 viewType을 다르게 반환하게 진행했습니다.
```kotlin
private var profileList = emptyList<HomeProfileModel>()
```
```kotlin
override fun getItemViewType(position: Int): Int {
     return when (profileList[position]) {
         is ProfileHeader -> HomeViewTypeModel.Header.type
         is ProfileBirthday -> HomeViewTypeModel.Birthday.type
         is Profile -> HomeViewTypeModel.Profile.type
     }
}
```

kez-lab replied:
음 제 질문의 의도는 왜 sealed class HomeProfileModel {
블럭 내에서 data class를 정의하지 않고 분리하셨는지 궁금했습니다!
}

Jokwanhee replied:
아.. 고건 제가 수정해야했는데, 수정을 안했네요.. 원래 sealed 클래스를 정의하지 않고 데이터 클래스를 먼저 정의하면서 시작해서 수정해야했는데 까먹고 있었습니다..ㅎ_ㅎ 수정할게요!

[e90d735](https://github.com/DO-SOPT-ANDROID/kwanhee-jo/pull/6/commits/e90d7350f623b999ca447474a1e6f030c786cf70)

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/presentation/home/MyPageFragment.kt:
```diff
@@ -0,0 +1,50 @@
+
+    private fun logout() {
+        binding.btnLogout.setOnClickListener {
+            activity?.apply {
+                getSharedPreferenceUser().edit(commit = true) {
+                    putBoolean(AUTO_LOGIN, false)
+                }
+                finish()
+            }
```
scope함수의 용도를 살짝 공부해보면 좋을 것 같습니다, apply는 주로 초기화 및 객체반환시 사용되며, 참조를 위해서 사용하시는거라면, run을 쓰는게 더 좋을 것 같네요

Jokwanhee replied:
@KwakEuiJin 엇! 반대 아닌가요? "apply ~ 객체 반환시 사용" 이라는 말이 어떤 의미인건가요?

kez-lab replied:
``` kotlin
val intent = Intent().apply {
    putExtra("key", value)
}
```

apply는 이런식으로 Intent 자체를 반환하기에 해당 객체를 초기화 하는 용도로 자주 사용됩니다.

Jokwanhee replied:
와 진짜 몰랐어요.. 스코프를 제대로 알고 사용한 것이 아니였다는 제 자신이 부끄럽네요.. 또 배워갑니다 갓의진!!!!
[444cada](https://github.com/DO-SOPT-ANDROID/kwanhee-jo/pull/6/commits/444cada1c5c9ae439cf6436483f26e29c36a4e53)

kez-lab commented on app/build.gradle:
```diff
@@ -26,6 +26,7 @@ android {
     }
     buildFeatures {
         viewBinding true
+        dataBinding true
     }
     compileOptions {
```
이번 과제에 dataBinding을 적용하신 이유가 있을까요??

librarywon replied:
이게 의진이 눈에 들어와버렸네

Jokwanhee replied:
BaseFragment 를 만들기 위해서 DataBindingUtil 클래스를 사용하기 위해서 데이터 바인딩을 사용했습니다.. 뷰 바인딩 으로 view를 인플레이트하기 위해서는 여러 인자가 필요해서 기존 `::inflate` 를 사용하는 방법으로는 해결되지 않아.. 굳이 필요없는 데이터 바인딩이 사용하게 되었습니다!

kez-lab replied:
오히려 데이터 바인딩이 layoutId가 필요하여 더 많은 인자가 필요하지 않나요?!! (단순 궁금)

Jokwanhee replied:
아 제가 말을 잘못했네요.. 말씀이 맞습니다!

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeAdapter.kt:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    fun setProfileList(profileList: List<HomeProfileModel>) {
+        this.profileList = profileList.toList()
+        notifyDataSetChanged()
```
👍

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/model/Profile.kt:
```diff
@@ -0,0 +1,19 @@
+    val birth: Long,
+    val update: Long,
+    val music: Music?
+): HomeProfileModel() {
+    fun checkBirthDay(): Boolean = SimpleDateFormat("yyyy-MM-dd").run {
+        format(System.currentTimeMillis()) == format(birth)
+    }
+
+    fun currentUpdateProfile(): Boolean =
+        System.currentTimeMillis() - update <= (1000 * 60 * 60 * 24)
+}
```
모델 내에서 이런식으로 간단한 계산 처리하는 거 너무 좋은 것 같아요

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/presentation/home/MyPageFragment.kt:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private fun initMyData() {
+        with(activity?.getSharedPreferenceUser()) {
+            binding.tvIdContent.text = this?.getString(ID, "") ?: return@with
```
```suggestion
        activity?.getSharedPreferenceUser()?.run {
            binding.tvIdContent.text = this?.getString(ID, "")
```

Jokwanhee replied:
[ef09303](https://github.com/DO-SOPT-ANDROID/kwanhee-jo/pull/6/commits/ef0930311c9f41c2a5e99f76ffcbbe74df780710)

오.. 감사합니다. 저도 불편한 코드였는데 깔끔하네요!

kez-lab reviewed (changes requested):
정말 빠르게 과제를 하셨군요!! 역시 👍 리뷰 확인 및 수정 부탁드립니다~!!

Jokwanhee committed:
[add] 상호님 임시 리스트 파일 피드백

테스트 더미 리스트 데이터를 따로 파일로 빼서 사용하는것을 말씀해주셔서, 피드백 수용!

Jokwanhee committed:
[add] 태희님 뷰 타입 sealed 클래스 적용

기존 companion object 사용말고 sealed 클래스를 사용하여서 view type 을 적용해봤다.

Jokwanhee committed:
[feedback] 의진님 스코프 함수 수정

기존 with 스코프 사용말고 run 스코프를 사용하여 널 허용으로 접근하면 return 반환을 사용하지 않는 유연한 코드를 작성할 수 있음을 알 수 있었다.

Jokwanhee committed:
[feedback] 의진님 Glide 확장함수 만들기

반복되는 코드의 양을 확장함수를 사용해서 많이 줄일 수 있었다.

Jokwanhee committed:
[mod] modify ArrayDeque to MutableList

프래그먼트가 다시 그려지는 과정에서 큐 자료구조의 생일인 친구를 계속 추가하는 것이 반복되어서 다시 그려질 때 새로운 리스트를 만들 수 있는 MutableList를 사용하여 새 참조 값을 부여하였다.

Jokwanhee committed:
[feedback] 의진님 데이터 클래스 분리의 이유

나누어진 데이터 클래스를 sealed 클래스 내부 정의로 바꾸었다.

Jokwanhee committed:
[feedback] 의진님 apply를 run 으로 수정

apply는 자신의 객체를 반환, run은 Unit을 반환하는 것을 알 수 있음.

Jokwanhee committed:
[feedback] 재원님 bias 삭제

별로 필요없는 bias는 제거해도 될듯하다!

Jokwanhee committed:
[add] bottom navigation reselected

바텀네비게이션을 클릭 후 다시 클릭하면 나타나는 리스너로 recyclerview scroll position 을 상단으로 올리는 기능

Jokwanhee committed:
[add] bottom navigation ui

바텀 네비게이션 ui 및 기능 추가

Jokwanhee committed:
[add] migration RecyclerView to ListAdatper & DiffUtil

리사이클러뷰에서 DiffUtil 로 마이그레이션하기

Jokwanhee committed:
[add] week2 notion into README.md

2주차 심화 과제 노션 링크

emjayMJkim commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeViewHolder.kt:
```diff
@@ -0,0 +1,40 @@
+            binding.tvStatusMessage.text = item.music.musicAlbumTitle()
+            binding.tvStatusMessage.visibility = View.VISIBLE
+        } else {
+            binding.tvStatusMessage.visibility = View.GONE
```
이 부분 XML에서 처리해줄 수도 있습니다!!

Jokwanhee replied:
@emjayMJkim 
헉.. 어떻게 하면 할 수 있는거죠? 데이터바인딩을 사용하는 부분인건가요!?

emjayMJkim replied:
넵 맞습니당~ (그냥 또 하나의 방법일 뿐입니다...ㅎㅎ)

emjayMJkim commented on app/src/main/java/org/sopt/dosopttemplate/model/HomeProfileModel.kt:
```diff
@@ -0,0 +1,33 @@
+        }
+
+        fun currentUpdateProfile(): Boolean =
+            System.currentTimeMillis() - update <= (1000 * 60 * 60 * 24)
```
멋있습니다..!

emjayMJkim reviewed (commented):
코드 진짜 멋있습니다.. 잘 보고 갑니당

Jokwanhee committed:
[add] landscape screen

가로모드 대응화면 만들기(viewPage2 & TabLayout) 사용
별도로 가로모드 Adapter 클래스 작성

Jokwanhee committed:
[add] adapter item click listener

클릭 이벤트 추가

Jokwanhee committed:
[add] bind HomeDetailActivity UI

상세 보기 화면 UI 바인딩

Jokwanhee committed:
[add] HomeDetail birthday & music UI

생일 & 뮤직 ui 추가

Jokwanhee committed:
[add] HomeDetail back button

세부 사항 액티비티 뒤로가기 버튼 만들기

Jokwanhee committed:
[add] add item & remove item

아이템 삭제 및 추가 (다이어로그 사용)
친구 생성 시 생일 확인 후 생일 프로필 생성
가로모드 대응
삭제 및 추가 시 리스트 갱신
친구 삭제 시 생일 + 기존 아이템 중복 삭제

yeonjeen commented on app/src/main/java/org/sopt/dosopttemplate/adapter/HomeViewHolder.kt:
```diff
@@ -0,0 +1,40 @@
+        if (item.profileImage == null) {
+            binding.ivUser.roundedCornerGlide(binding.root, R.drawable.kakao_profile, 150, 50)
+        } else {
+            binding.ivUser.roundedCornerGlide(binding.root, item.profileImage, 150, 50)
```
roundedCornerGlide를 사용하는 큰 이점이 뭘까요?

Jokwanhee replied:
확장함수를 사용한 함수입니다! 코틀린은 자바와 다르게 최상위 함수를 가집니다. 자바는 클래스로 묶어서 다니죠.. 재사용성이랑 코드의 가독성측면이라는 이점이 있다고 생각합니다. 추가로 코틀린스러운 코드라고 생각합니당

yeonjeen commented on app/src/main/java/org/sopt/dosopttemplate/presentation/home/MyPageFragment.kt:
```diff
@@ -0,0 +1,50 @@
+import org.sopt.dosopttemplate.presentation.LoginActivity.Companion.NICKNAME
+import org.sopt.dosopttemplate.presentation.LoginActivity.Companion.getSharedPreferenceUser
+
+class MyPageFragment : BaseFragment<FragmentMypageBinding>() {
```
관희님 코드 보면서 BaseFragment/BaseActivity가 있다는 걸 알았어요!! 많이 배우고 갑니다

yeonjeen commented on app/src/main/java/org/sopt/dosopttemplate/util/SampleList.kt:
```diff
@@ -0,0 +1,129 @@
+            R.drawable.raichu,
+            1697565421000,
+            1697565421000,
+            Music("노래 제목이 길게하면 어떨까?", "박효신")
```
박효신 엄청 좋아하시네요 ㅋㅋ

yeonjeen reviewed (commented):
항상 과제 1등!! 멋있어요!! 수고하셨습니다~

Jokwanhee committed:
[review] 의진 리뷰

프로필 데이터를 생일 데이터로 매핑하는 확장함수 만들기

Jokwanhee committed:
[review] 의진 리뷰 피드백

어뎁터 사용으로 메모리 누수 방지
- RecyclerView adapter
- ViewPager2 adapter

Jokwanhee committed:
[del] Extensions func about Glide

쓸모없는 Glide 확장 함수 코드 삭제

Jokwanhee committed:
Merge pull request #8 from DO-SOPT-ANDROID/feat/week2_challenge

PR : 실습 2주차 과제

Jokwanhee committed:
Merge pull request #7 from DO-SOPT-ANDROID/feat/week2_advanced

PR : 실습 2주차 과제
